rust-runtime/aws-smithy-cbor/src/decode.rs

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use std::borrow::Cow;
+
+use aws_smithy_types::{Blob, DateTime};
+use minicbor::decode::Error;
+
+use crate::data::Type;
+
+/// Provides functions for decoding a CBOR object with a known schema.
+///
+/// Although CBOR is a self-describing format, this decoder is tailored for cases where the schema
+/// is known in advance. Therefore, the caller can determine which object key exists at the current
+/// position by calling `str` method, and call the relevant function based on the predetermined schema
+/// for that key. If an unexpected key is encountered, the caller can use the `skip` method to skip
+/// over the element.
+#[derive(Debug, Clone)]
+pub struct Decoder<'b> {
+    decoder: minicbor::Decoder<'b>,
+}
+
+/// When any of the decode methods are called they look for that particular data type at the current
+/// position. If the CBOR data tag does not match the type, a `DeserializeError` is returned.
+#[derive(Debug)]
+pub struct DeserializeError {
+    #[allow(dead_code)]
+    _inner: Error,
+}
+
+impl std::fmt::Display for DeserializeError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self._inner.fmt(f)
+    }
+}
+
+impl std::error::Error for DeserializeError {}
```

Comment:
Don't want to expose `minicbor` in our public API. But `source` doesn't really matter for our errors, since we're not doing and won't be doing any error reporting walking an error chain.